dkg/dkg_test.go

```diff
@@ -197,44 +248,147 @@ func testDKG(t *testing.T, def cluster.Definition, p2pKeys []*ecdsa.PrivateKey)
 	}
 }
 
-// startBootnode starts a charon bootnode and returns its http ENR endpoint.
-func startBootnode(ctx context.Context, t *testing.T) (string, <-chan error) {
+func TestSyncFlow(t *testing.T) {
+	if !*slow {
+		t.Skip("skipping slow tests")
+	}
+
+	const (
+		nodes = 4
+		vals  = 2
+	)
+
+	lock, keys, _ := cluster.NewForT(t, vals, nodes, nodes, 0)
+
+	tests := []struct {
+		name       string
+		connect    []int // Initial connections
+		disconnect []int // Drop some peers
+		reconnect  []int // Connect remaining peers
+	}{
+		{
+			name:       "first",
+			connect:    []int{0, 1, 2},
+			disconnect: []int{0, 1},
+			reconnect:  []int{0, 1, 3},
+		},
+		{
+			name:       "second",
+			connect:    []int{0, 1, 3},
+			disconnect: []int{3},
+			reconnect:  []int{2, 3},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			// Start bootnode
+			bootnode, errChan := startBootnode(ctx, t)
+
+			dir, err := os.MkdirTemp("", "")
+			require.NoError(t, err)
+
+			configs := getConfigs(t, lock.Definition, keys, dir, bootnode)
+
+			dkgs := make([]*testDkg, nodes)
+
+			// Start DKG for initial peers.
+			for _, idx := range test.connect {
+				dkgs[idx] = startNewDKG(configs[idx], errChan)
+			}
+
+			// Drop some peers.
+			time.Sleep(time.Second * 5) // Give some time to connect.
```

Comment:
not a fan of sleeping. Since it is brittle and will flap. Rather wait for callbacks.